In [1]:
from sklearn.datasets import load_diabetes
from sklearn import tree
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn import linear_model
from sklearn.dummy import DummyRegressor
from sklearn.metrics import r2_score
import pandas as pd

1. В документации библиотеки scikit learn найдите, как загрузить датасет для построения модели прогноза прогрессирования заболевания через год после исходного уровня.

In [2]:
my_datasets = load_diabetes()

2. Выведите содержимое поля DESCR, которое вернет функция, загружающая датасет, чтобы изучить содержимое датасета.

In [3]:
print(my_datasets.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

3. Подготовьте данные к обучению моделей: отделите целевой признак от датасета, разделите датасет на тренировочную и валидационную части.

In [4]:
df = pd.DataFrame(my_datasets.data, columns=my_datasets.feature_names)
df['target'] = my_datasets.target
features = df.drop(['target'], axis=1)
#Отделяем целевой признак от датасета
target = df['target']
#Разделяем датасет на тренировочную и валидационную части
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.3, random_state=45)

4. Обучите решающее дерево и оцените адекватность обученной модели. Для этого достаточно сравнить метрики качества с метриками, получаемыми от простой модели. Например, если бы мы подавали на вход среднее значение целевой переменной на обучающей части выборки.

In [5]:
#Обучаем решающее дерево
model = tree.DecisionTreeRegressor()
model.fit(features_train, target_train)
#Предсказываем значения целевой переменной на основе тестовой выборки
predictions_val = model.predict(features_test)
print(r2_score(target_test, predictions_val))

-0.24486792864226992


In [6]:
#Оценим адекватность модели путем сравнения метрики R2 нашей модели с метрикой R2 dummyRegressorModel
dummyRegressorModel = DummyRegressor(strategy="mean")
dummyRegressorModel.fit(features_train, target_train)
predictions_val_dummyRegressorModel = dummyRegressorModel.predict(features_test)
print(r2_score(target_test, predictions_val_dummyRegressorModel))

-0.003102686796445875


Вывод:
Из полученных результатов мы видим, что метрика R2 модели класса DecisionTreeRegressor (-0.244) меньше, чем метрика R2 модели класса DummyRegressor (-0.003), что говорит о неадекватности полученной модели.

5. Обучите линейную регрессию и оцените ее адекватность.

In [7]:
#Обучаем линейную регрессию
linerRegresModel = linear_model.LinearRegression()
linerRegresModel.fit(features_train, target_train)
#Предсказываем значения целевой переменной на основе тестовой выборки
predictLinerRegresVal = linerRegresModel.predict(features_test)
#Оценим адекватность модели по метрике R2
print(r2_score(target_test, predictLinerRegresVal))

0.5326715876099559


Вывод: Полученная модель имеет достаточно неплохое качество. Значение 0.53 говорит о том, что модель работает лучше среднего значения и лучше модели класса DummyRegressor.

6. Выберите лучшую модель и обоснуйте свой выбор.

Датасет содержит количественные признаки. Для таких признаков лучше всего подходят регрессионные модели, в чем мы и убедились на практике.